File: src/lib/charts/muscleAggregate.ts
/**
 * The aggregation behind FIVE charts — Muscle Balance, Volume by Body Part,
 * Muscle Freshness, Volume Stream and the Muscle Contour Map. PURE, so the
 * arithmetic can be tested without a database.
 *
 * It lived inside `useMuscleAnalytics`' queryFn, which meant the only way to
 * check any of it was to look at the rendered chart and decide whether the
 * numbers felt right.
 */

/** Canonicalize Hevy muscle tags into the 6 display groups (v5.1 aliases included). */
export const MUSCLE_MAP: Record<string, string> = {
  chest: 'Chest', pecs: 'Chest',
  back: 'Back', lats: 'Back', traps: 'Back', rhomboids: 'Back', 'upper back': 'Back', 'lower back': 'Back',
  shoulders: 'Shoulders', delts: 'Shoulders', rear_delts: 'Shoulders', side_delts: 'Shoulders', front_delts: 'Shoulders',
  biceps: 'Arms', triceps: 'Arms', forearms: 'Arms', arms: 'Arms',
  // `adductors` / `inner_thigh` were both absent, so the hip-adduction machine
  // credited no display group at all and the Freshness map showed the legs as
  // fresher than they were.
  quads: 'Legs', quadriceps: 'Legs', hamstrings: 'Legs', glutes: 'Legs', calves: 'Legs',
  abductors: 'Legs', adductors: 'Legs', inner_thigh: 'Legs', legs: 'Legs',
  core: 'Core', abs: 'Core', abdominals: 'Core', obliques: 'Core',
}

export const MUSCLE_GROUPS = ['Chest', 'Back', 'Shoulders', 'Arms', 'Legs', 'Core'] as const

export interface MuscleSetRow {
  /** Row identity — falls back for the dedupe key when there is no pair. */
  id: string
  weightKg: number
  reps: number
  /** Unilateral L/R rows share this. Two rows, ONE set. */
  pairId: string | null
  /** Which limb this row logged. A `pairId` only collapses WITH a side. */
  side?: 'L' | 'R' | null
  /** Display groups this row credits (already canonicalised, deduped). */
  groups: string[]
  /** Session date, YYYY-MM-DD. */
  date: string
}

export interface MuscleStat { group: string; sets: number; volume: number; daysSince: number | null }

export interface MuscleAggregate {
  stats: MuscleStat[]
  weekly: Array<Record<string, number | string>>
}

/**
 * Per-row tonnage with unilateral pairs collapsed to the WEAKER side.
 *
 * `sessionVolumeKg` cannot be called directly here: this aggregator credits each
 * row to its own muscle groups, so it needs a per-ROW number, not a session
 * total. The two must still agree, so a genuine L/R pair splits its collapsed
 * total in half across its two rows — the pair then sums to exactly
 * `2 × min(weight) × min(reps)`, which is what `sessionVolumeKg` returns, while
 * each row keeps its own attribution.
 *
 * Rule 1 in this file already deduped the SET COUNT for pairs. Tonnage was never
 * deduped, so "L 5 kg × 10, R 5 kg × 14" credited 120 kg to the group here and
 * 100 kg on the session card — the same work, two answers, and the drift grew
 * with every asymmetric week.
 *
 * A lone side, and a malformed bucket of 3+ rows, score as logged — the same
 * fallbacks `sessionVolumeKg` takes.
 */
function effectiveVolumes(rows: readonly MuscleSetRow[]): Map<string, number> {
  const out = new Map<string, number>()
  const pairs = new Map<string, MuscleSetRow[]>()
  const raw = (r: MuscleSetRow) => (r.weightKg || 0) * (r.reps || 0)

  for (const r of rows) {
    if (r.pairId && (r.side === 'L' || r.side === 'R')) {
      const bucket = pairs.get(r.pairId) ?? []
      bucket.push(r)
      pairs.set(r.pairId, bucket)
      continue
    }
    out.set(r.id, raw(r))
  }

  for (const bucket of pairs.values()) {
    const left = bucket.find((x) => x.side === 'L')
    const right = bucket.find((x) => x.side === 'R')
    if (bucket.length === 2 && left && right) {
      // ONE set's tonnage, at the weaker side, landing on ONE row — the second
      // row of the pair contributes nothing. Splitting it across both rows and
      // letting the caller sum them credited the pair TWICE (2026-08-18): a
      // split single-arm raise outweighed the identical set logged unsided.
      // `sessionVolumeKg` carries the full argument.
      const one = Math.min(left.weightKg || 0, right.weightKg || 0)
        * Math.min(left.reps || 0, right.reps || 0)
      out.set(right.id, one)
      out.set(left.id, 0)
    } else {
      for (const x of bucket) out.set(x.id, raw(x))
    }
  }
  return out
}

/** Sunday-anchored week start for a YYYY-MM-DD date. */
export function weekStartUTC(dateISO: string): string {
  const d = new Date(`${dateISO}T00:00:00Z`)
  d.setUTCDate(d.getUTCDate() - d.getUTCDay())
  return d.toISOString().slice(0, 10)
}

/**
 * Fold rows into per-group set counts, tonnage and recency, plus a per-week
 * set series.
 *
 * TWO RULES THAT ARE EASY TO GET WRONG, both load-bearing:
 *
 * 1. A unilateral exercise logs left and right as separate rows sharing a
 *    `pairId`. That is ONE set for every set-count purpose (balance radar,
 *    freshness, contour map, volume stream) and TWO rows for tonnage. Counting
 *    it as two sets doubled every cable lateral raise in the volume landmarks.
 * 2. A row tagged `quads + glutes` credits a set to BOTH — but only once each,
 *    even if the same pair appears twice in the input.
 */
export function aggregateMuscleSets(rows: readonly MuscleSetRow[], todayISO: string): MuscleAggregate {
  const agg = new Map<string, { sets: number; volume: number; last: string | null }>()
  const weekMap = new Map<string, Record<string, number>>()
  const countedSets = new Set<string>()     // `${group}|${dedupeKey}`
  const countedWeekly = new Set<string>()   // `${week}|${group}|${dedupeKey}`
  const effective = effectiveVolumes(rows)

  for (const r of rows) {
    const groups = new Set(r.groups)
    if (!groups.size) continue
    const week = weekStartUTC(r.date)
    const dedupeKey = r.pairId ?? r.id
    const vol = effective.get(r.id) ?? (r.weightKg || 0) * (r.reps || 0)
    for (const g of groups) {
      const a = agg.get(g) ?? { sets: 0, volume: 0, last: null }
      a.volume += vol
      if (!a.last || r.date > a.last) a.last = r.date
      const setKey = `${g}|${dedupeKey}`
      if (!countedSets.has(setKey)) { countedSets.add(setKey); a.sets += 1 }
      agg.set(g, a)
      const w = weekMap.get(week) ?? {}
      const weekKey = `${week}|${g}|${dedupeKey}`
      if (!countedWeekly.has(weekKey)) { countedWeekly.add(weekKey); w[g] = (w[g] ?? 0) + 1 }
      weekMap.set(week, w)
    }
  }

  const todayMs = Date.parse(`${todayISO}T00:00:00Z`)
  const stats: MuscleStat[] = MUSCLE_GROUPS.map((g) => {
    const a = agg.get(g)
    const daysSince = a?.last
      ? Math.round((todayMs - Date.parse(`${a.last}T00:00:00Z`)) / 86_400_000)
      : null
    return { group: g, sets: a?.sets ?? 0, volume: Math.round(a?.volume ?? 0), daysSince }
  })

  const weekly = [...weekMap.entries()]
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([week, groups]) => ({
      week: week.slice(5),
      ...Object.fromEntries(MUSCLE_GROUPS.map((g) => [g, groups[g] ?? 0])),
    }))

  return { stats, weekly }
}

File: native/Packages/OnyxCore/Sources/OnyxCore/Charts/MuscleAggregate.swift
import Foundation

/// The aggregation behind the Muscle Balance / Volume by Body Part / Freshness /
/// Volume Stream charts — a port of `src/lib/charts/muscleAggregate.ts`.
///
/// Two load-bearing rules: a unilateral pair (shared `pairId` AND sides L+R,
/// exactly two rows) is ONE set and ONE tonnage at the weaker side; a row tagged
/// with the same group twice credits it once.
public struct MuscleSetRow: Codable, Sendable, Equatable {
    public var id: String
    public var weightKg: Double
    public var reps: Double
    public var pairId: String?
    public var side: String?
    public var groups: [String]
    public var date: String

    public init(id: String, weightKg: Double, reps: Double, pairId: String?, side: String? = nil, groups: [String], date: String) {
        self.id = id; self.weightKg = weightKg; self.reps = reps; self.pairId = pairId; self.side = side; self.groups = groups; self.date = date
    }
}

public struct MuscleStat: Codable, Sendable, Equatable {
    public var group: String
    public var sets: Int
    public var volume: Double
    public var daysSince: Int?
}

/// `{ week: "MM-DD", Chest: n, Back: n, … }` — a mixed-value JS object.
public struct WeeklyMuscleRow: Sendable, Equatable {
    public var week: String
    public var counts: [(String, Int)]

    public static func == (a: WeeklyMuscleRow, b: WeeklyMuscleRow) -> Bool {
        a.week == b.week && a.counts.count == b.counts.count && zip(a.counts, b.counts).allSatisfy { $0.0 == $1.0 && $0.1 == $1.1 }
    }
}

public struct MuscleAggregate: Sendable, Equatable {
    public var stats: [MuscleStat]
    public var weekly: [WeeklyMuscleRow]
}

public enum MuscleAggregator {
    /// Hevy tag → one of six display groups, in table order.
    public static let map: [(String, String)] = [
        ("chest", "Chest"), ("pecs", "Chest"),
        ("back", "Back"), ("lats", "Back"), ("traps", "Back"), ("rhomboids", "Back"), ("upper back", "Back"), ("lower back", "Back"),
        ("shoulders", "Shoulders"), ("delts", "Shoulders"), ("rear_delts", "Shoulders"), ("side_delts", "Shoulders"), ("front_delts", "Shoulders"),
        ("biceps", "Arms"), ("triceps", "Arms"), ("forearms", "Arms"), ("arms", "Arms"),
        ("quads", "Legs"), ("quadriceps", "Legs"), ("hamstrings", "Legs"), ("glutes", "Legs"), ("calves", "Legs"),
        ("abductors", "Legs"), ("adductors", "Legs"), ("inner_thigh", "Legs"), ("legs", "Legs"),
        ("core", "Core"), ("abs", "Core"), ("abdominals", "Core"), ("obliques", "Core"),
    ]
    public static let groups = ["Chest", "Back", "Shoulders", "Arms", "Legs", "Core"]

    static func raw(_ r: MuscleSetRow) -> Double { (r.weightKg.isFinite ? r.weightKg : 0) * (r.reps.isFinite ? r.reps : 0) }

    /// Per-row tonnage with unilateral pairs collapsed to the WEAKER side, landing on the R row.
    static func effectiveVolumes(_ rows: [MuscleSetRow]) -> [String: Double] {
        var out: [String: Double] = [:]
        var pairOrder: [String] = []
        var pairs: [String: [MuscleSetRow]] = [:]
        for r in rows {
            if let p = r.pairId, !p.isEmpty, r.side == "L" || r.side == "R" {
                if pairs[p] == nil { pairOrder.append(p) }
                pairs[p, default: []].append(r)
                continue
            }
            out[r.id] = raw(r)
        }
        for p in pairOrder {
            let bucket = pairs[p]!
            let left = bucket.first { $0.side == "L" }
            let right = bucket.first { $0.side == "R" }
            if bucket.count == 2, let l = left, let r = right {
                let one = min(l.weightKg, r.weightKg) * min(l.reps, r.reps)
                out[r.id] = one
                out[l.id] = 0
            } else {
                for x in bucket { out[x.id] = raw(x) }
            }
        }
        return out
    }

    /// Sunday-anchored week start for a YYYY-MM-DD date.
    public static func weekStartUTC(_ dateISO: String) -> String {
        Swap.dateForWeekday(dateISO, 0)
    }

    public static func aggregate(_ rows: [MuscleSetRow], todayISO: String) -> MuscleAggregate {
        struct Agg { var sets = 0; var volume = 0.0; var last: String? = nil }
        var agg: [String: Agg] = [:]
        var weekOrder: [String] = []
        var weekMap: [String: [(String, Int)]] = [:]   // insertion-ordered per week
        var countedSets = Set<String>()
        var countedWeekly = Set<String>()
        let effective = effectiveVolumes(rows)

        for r in rows {
            // `new Set(r.groups)` — insertion order, duplicates dropped.
            var groups: [String] = []
            for g in r.groups where !groups.contains(g) { groups.append(g) }
            if groups.isEmpty { continue }
            let week = weekStartUTC(r.date)
            let dedupeKey = r.pairId ?? r.id
            let vol = effective[r.id] ?? raw(r)
            for g in groups {
                var a = agg[g] ?? Agg()
                a.volume += vol
                if a.last == nil || r.date > a.last! { a.last = r.date }
                let setKey = "\(g)|\(dedupeKey)"
                if !countedSets.contains(setKey) { countedSets.insert(setKey); a.sets += 1 }
                agg[g] = a
                if weekMap[week] == nil { weekOrder.append(week); weekMap[week] = [] }
                let weekKey = "\(week)|\(g)|\(dedupeKey)"
                if !countedWeekly.contains(weekKey) {
                    countedWeekly.insert(weekKey)
                    if let i = weekMap[week]!.firstIndex(where: { $0.0 == g }) { weekMap[week]![i].1 += 1 } else { weekMap[week]!.append((g, 1)) }
                }
            }
        }

        let todayN = ISODate.dayNumber(todayISO)
        let stats = groups.map { g -> MuscleStat in
            let a = agg[g]
            var daysSince: Int? = nil
            if let last = a?.last, let t = todayN, let l = ISODate.dayNumber(last) { daysSince = t - l }
            return MuscleStat(group: g, sets: a?.sets ?? 0, volume: jsRound(a?.volume ?? 0), daysSince: daysSince)
        }

        let weekly = weekOrder.sorted().map { week -> WeeklyMuscleRow in
            let counts = weekMap[week] ?? []
            return WeeklyMuscleRow(week: String(week.dropFirst(5)), counts: groups.map { g in (g, counts.first { $0.0 == g }?.1 ?? 0) })
        }
        return MuscleAggregate(stats: stats, weekly: weekly)
    }
}
